### PR TITLE
Use PyTorch CUTLASS for BF16 grouped GEMM on SM100

### PR DESCRIPTION
**Pull request by wujingyue**
# Description

Motivated by [NVIDIA/Megatron-LM#7400](https://github.com/NVIDIA/Megatron-LM/issues/7400): on GB200, grouped BF16 GEMM can retain old output values despite beta=0 while NCCL all-gather and reduce-scatter run concurrently. The issue includes a [standalone cuBLASLt + NCCL reproducer](https://github.com/NVIDIA/Megatron-LM/issues/7400#issuecomment-5720958195) without PyTorch or TE. The underlying cause remains unresolved.

This PR extends `NVTE_USE_CUTLASS_GROUPED_GEMM=1` to the graph-safe BF16 GroupedLinear path on SM100 using `torch._grouped_mm`. It provides an alternative to the affected cuBLASLt path while preserving independent communication streams and CUDA graph support.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Changes

- Dispatch supported BF16 forward, dgrad, and overwrite wgrad operations through PyTorch CUTLASS in the shared grouped GEMM wrapper. Expert counts remain on the GPU, and caller-owned output buffers and unused capacity are preserved.
- Retain TE's existing backend for FP32 outputs, fused accumulation, unsupported shapes, and explicit alpha/beta. Adding a rounded BF16 product separately would not preserve fused accumulation precision.
- Reuse the existing CUTLASS flag and extend its existing accuracy test to SM100. The module API also requires `NVTE_GROUPED_LINEAR_USE_FUSED_GROUPED_GEMM=1`; Hopper keeps its existing dispatch. If PyTorch exposes `torch.backends.cuda.matmul.prefer_cublaslt_grouped_gemm`, it must be `False`.

The implementation uses temporary weight/output buffers. No performance improvement is claimed.

## Validation

GPU validation used GB200, PyTorch `2.14.0a0+4fdf77b940.nv26.08`, and TE `2.18.0+27486e03`:

- **241 existing tests passed, 19 skipped**, covering forward/dgrad/wgrad, bias, empty experts, caller buffers, deferred wgrad, accumulation and MXFP8 fallbacks, and CUDA graph capture/replay. Instrumentation confirmed all three layouts reached `_grouped_mm`.
- **8-GPU MFSDP v2 + Flex/HybridEP training:** 100 eager steps and 100 steps with both training and optimizer CUDA graphs completed with finite gradient norms. Training validation used an earlier revision of the adapter; the existing tests above were rerun after the API cleanup and implementation simplification.
- Isort, Black, Ruff F checks, and `git diff --check` passed.

This draft ports the change onto `main`. The new GEMM helper and shared dispatch function are identical to the GPU-tested versions; the port only required resolving imports. A fresh native build/test on this `main` revision and a physical Hopper run remain outstanding.

# Checklist

- [x] Read the contributing guidelines
- [x] Documented the flag and backend limitations
- [x] Extended existing test coverage
- [x] Ran targeted numerical and CUDA graph tests on GB200
- [ ] Validate a fresh build of the PR revision
- [ ] Run the existing CUTLASS tests on Hopper

### CODE REVIEW

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 1]
```diff
@@ -638,6 +705,19 @@ def _is_fp8_blockwise(operand) -> bool:
     if bias_scale is not None and bias is None:
         raise ValueError("bias_scale requires bias to be provided.")
 
+    # PyTorch returns a BF16 product, so a separate add would lose the precision
+    # of TE's fused accumulation. Leave that case on the existing backend.
+    if (
+        os.getenv("NVTE_USE_CUTLASS_GROUPED_GEMM", "0") == "1"
+        and hasattr(torch, "_grouped_mm")
```
Comment (wujingyue) [inline transformer_engine/pytorch/cpp_extensions/gemm.py]: Really? Torch has that for a long time.

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 2]
```diff
@@ -638,6 +705,19 @@ def _is_fp8_blockwise(operand) -> bool:
...
+        and torch.cuda.get_device_capability() == (10, 0)
+        and not accumulate
+        and alpha is None
+        and beta is None
+        and _pytorch_grouped_gemm(A, B, out, layout=layout, bias=bias, bias_scale=bias_scale)
```
Comment (wujingyue) [inline transformer_engine/pytorch/cpp_extensions/gemm.py]: I don't like this. Some preconditions are outside of _pytorch_grouped_gemm and some are inside. We ought to unify them in one place.

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 3]
```diff
@@ -566,6 +569,81 @@ def _get_grouped_cublas_workspace(device: int, layout: str) -> torch.Tensor:
...
+    outputs = out if isinstance(out, list) else [out]
+    for tensor in [B, *inputs, *outputs]:
+        if isinstance(tensor, GroupedTensorStorage):
+            tensor = tensor.rowwise_data
+        if (
+            tensor is None
+            or isinstance(tensor, QuantizedTensorStorage)
+            or tensor.dtype != torch.bfloat16
+        ):
+            return False
```
Comment (wujingyue) [inline transformer_engine/pytorch/cpp_extensions/gemm.py]: This is a bit convoluted.

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 4]
```diff
@@ -566,6 +569,81 @@ def _get_grouped_cublas_workspace(device: int, layout: str) -> torch.Tensor:
...
+    if not B.all_same_last_dim() or (layout != "NT" and isinstance(out, list)):
+        return False
+    if isinstance(A, list):
+        shapes = [tensor.shape[-1:] if layout == "NT" else tensor.shape for tensor in A]
+        if any(shape != shapes[0] for shape in shapes):
+            return False
+    elif not A.all_same_last_dim() or (layout != "NT" and not A.all_same_shape()):
+        return False
+    if getattr(torch.backends.cuda.matmul, "prefer_cublaslt_grouped_gemm", False):
```
Comment (wujingyue) [inline transformer_engine/pytorch/cpp_extensions/gemm.py:614]: Is getattr necessary?

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 5]
```diff
@@ -566,6 +569,81 @@ def _get_grouped_cublas_workspace(device: int, layout: str) -> torch.Tensor:
...
+        offsets = torch.cumsum(B.first_dims, dim=0, dtype=torch.int32)
+
+    result = torch._grouped_mm(
+        b.T if layout == "NT" else b, a.mT if layout == "TN" else a, offs=offsets
+    )
+    rows = torch.arange(b.size(0), device=b.device) if layout != "NT" else None
+    if bias is not None:
+        bias_data = bias.rowwise_data.view(groups, 1, -1).float()
+        if rows is not None:
+            group_ids = torch.bucketize(rows, offsets, right=True).clamp_max(groups - 1)
```
Comment (wujingyue) [inline transformer_engine/pytorch/cpp_extensions/gemm.py]: What's this clamp_max doing and why?

File: transformer_engine/pytorch/cpp_extensions/gemm.py [hunk 6]
```diff
@@ -566,6 +569,94 @@ def _get_grouped_cublas_workspace(device: int, layout: str) -> torch.Tensor:
...
+            "PyTorch CUTLASS grouped GEMM does not support accumulation or explicit alpha/beta."
+        )
+
+    inputs = A if isinstance(A, list) else [A]
+    outputs = out if isinstance(out, list) else [out]
+    for tensor in [B, *inputs, *outputs]:
+        # Quantized wrappers can report a logical BF16 dtype.
+        if isinstance(tensor, QuantizedTensorStorage):
+            raise NotImplementedError("Quantized inputs and outputs are not supported.")
+        if isinstance(tensor, GroupedTensorStorage):
```
Comment (wujingyue) [inline transformer_engine/pytorch/cpp_extensions/gemm.py:600]: This looks silly -- why is GroupedTensorStorage used for non-quantized where the only field people care is rowwise_data?
